bin/cheribsdtest/cheribsdtest_cheriabi.c

```diff
@@ -143,21 +145,42 @@ struct adjacent_mappings {
 };
 
 /*
- * "Reserve" a chunk of address space by mapping pages, unmapping them,
- * and returning the base address.  This is necessary because small
- * mappings (e.g. 2 pages) might land between other mappings and thus
- * the next N pages might not be free.
+ * Find a region of address space with nothing mapped in it.
  */
 static ptraddr_t
 reserve_address_space(size_t len)
 {
-	void *addr;
-
-	addr = CHERIBSDTEST_CHECK_SYSCALL(
-	    mmap(0, len, PROT_READ | PROT_WRITE, MAP_ANON, -1, 0));
-	CHERIBSDTEST_CHECK_SYSCALL(munmap(addr, len));
-	/* XXX: With temporal safety we will need to force revocation here. */
-	return ((ptraddr_t)addr);
+	struct procstat *psp;
+	struct kinfo_proc *kipp;
+	struct kinfo_vmentry *kivp;
+	uint pcnt, vmcnt;
+	ptraddr_t addr = 0;
+
+	psp = procstat_open_sysctl();
+	CHERIBSDTEST_VERIFY(psp != NULL);
+	kipp = procstat_getprocs(psp, KERN_PROC_PID, getpid(), &pcnt);
+	CHERIBSDTEST_VERIFY(kipp != NULL);
+	CHERIBSDTEST_VERIFY(pcnt == 1);
+	kivp = procstat_getvmmap(psp, kipp, &vmcnt);
+	CHERIBSDTEST_VERIFY(kivp != NULL);
+
+	for (u_int i = 1; i < vmcnt; i++) {
+		if (kivp[i].kve_start - kivp[i-1].kve_end)
+			addr = kivp[i-1].kve_end;
+	}
+	/*
+	 * XXX: we've ignored possible gaps at the top and bottom of the
+	 * address space...
```

Comment:
I'm not sure what you be by "the small bit below `.text`". There's an oddity that you can't generally map the first page of the address space and there's no entry associated with that, but we do map things below the program and rtld.

Reply:
Oh, for some reason I thought we loaded programs fairly low in the address space.  Either way, it's not that we can't map things there, it's that there will be an entry "towards the bottom" of the address space, so skipping the interval between 0 and that is probably fine.

Reply:
We might be, I can't remember what our default address is. IIRC we load legacy programs a bit above 2^32 on 64-bit platforms. That's still fairly small in the grand scheme of things (If you're otherwise out of VA in cheribsdtest things have gotten quite weird).